### ISSUE TITLE
Connecting UEFI drivers at TPL_CALLBACK level causes IP assignment issue during BIOS POST (Bugzilla Bug 3935)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2-platforms#667]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 3935
Date: 2022-05-27T10:46:47+00:00
From: Dhanaraj V &lt;<vdhanaraj>&gt;
To:   unassigned &lt;<unassigned>&gt;
CC:   @ChaselChiu, @lgao4, maciej.rabeda, @nate-desimone, vdhanaraj

Last updated: 2022-11-01T21:50:31+00:00

**Comment by tianocore-issues** [comment]
## Comment 18976
Date: 2022-05-27 10:46:47 +0000
From: Dhanaraj V &lt;<vdhanaraj>&gt;

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): unassigned &lt;<unassigned>&gt;

Created attachment 986
Fix to connect UEFI drivers at TPL_APPLICATION level

Platform: Eaglestream 

Issue: IP is not assigned to SUT during BIOS POST, even though DHCP server is connected. 

Root Cause:

Ip6NdFasterTimerTicking() timer callback function is not executed during BIOS POST. Timer callback function Ip6NdFasterTimerTicking() is registered for TPL_CALLBACK, whereas the current TPL during its registration is also TPL_CALLBACK. Hence, Ip6NdFasterTimerTicking() does not get control during BIOS POST. This is expected behaviour as per CreateEvent() definition of UEFI specification,

&quot;If the current TPL is equal to or greater than the queued notification, it will wait until the TPL is lowered via EFI_BOOT_SERVICES.RestoreTPL().&quot;

Why Ip6NdFasterTimerTicking() is registered at TPL_CALLBACK level? It is found that BdsAfterConsoleReadyBeforeBootOptionCallback() is executing at level TPL_CALLBACK. Thereby all UEFI drivers are started at TPL_CALLBACK level. Hence, Ip6CreateService() which registers timer event callback for Ip6NdFasterTimerTicking() is also executed at TPL_CALLBACK.

In short, connecting UEFI drivers at TPL_CALLBACK level causes the issue.

Solution:

When the UEFI drivers are connected at TPL_APPLICATION level, the issue is solved. The attached modification moves the EfiBootManagerConnectAll() call from ConnectSequence() to PlatformBootManagerAfterConsole().

Attachment: [BZ-986-ConnectController_At_TPL_APP.7z](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2-platforms/BZ-986-ConnectController_At_TPL_APP.7z)

**Comment by tianocore-issues** [comment]
## Comment 18982
Date: 2022-05-31 21:40:45 +0000
From: @lgao4

EfiBootManagerConnectAll will break fast boot. Include package maintainter to discuss this issue.

**Comment by tianocore-issues** [comment]
## Comment 18983
Date: 2022-05-31 21:41:21 +0000
From: @lgao4

EfiBootManagerConnectAll will break fast boot. Include package maintainter to discuss this issue.

**Comment by tianocore-issues** [comment]
## Comment 19859
Date: 2022-10-28 02:10:36 +0000
From: Dhanaraj V &lt;<vdhanaraj>&gt;

Can you please provide a better solution to address this issue.

**Comment by tianocore-issues** [comment]
## Comment 19903
Date: 2022-11-01 21:50:31 +0000
From: @lgao4

Include NetworkPkg maintainer maciej.rabeda@intel.com also.
